File: pix_elixir/apps/lastfm/lib/lastfm.ex
defmodule Lastfm do
  @moduledoc """
  Word of the day
  """

  use GenServer
  use Tesla
  plug(Tesla.Middleware.BaseUrl, "https://ws.audioscrobbler.com/")
  plug(Tesla.Middleware.JSON)

  alias Display.Draw
  alias Display.Draw.Symbol

  @timeout 100
  @fetch_timeout 1000 * 60 * 2
  @offset 9
  @text_color 5
  @notes_color 6

  def start_link(_opts) do
    GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
  end

  def init(_state) do
    send(self(), :tick)
    send(self(), :fetch)

    Display.time(__MODULE__, 15000)

    state = %{
      text: "fetching... ",
      letter: 0,
      position: 0,
    }

    {:ok, state}
  end

  def terminate(_reason, state) do
    Display.remove(__MODULE__)

    {:ok, state}
  end

  def handle_info(:tick, state) do
    state = tick(state)

    data =
      Draw.empty()
      |> Draw.symbol({Symbol, "note"}, 1, 1, @notes_color)
      |> Draw.symbol({Symbol, "note"}, 8, 1, @notes_color)
      |> draw_text(state.text, @text_color, state.position, state.letter)

    Process.send_after(self(), :tick, @timeout)

    Display.update(__MODULE__, data)

    {:noreply, state}
  end

  def handle_info(:fetch, state) do
    Process.send_after(self(), :fetch, @fetch_timeout)
    fetch()
    {:noreply, state}
  end

  def handle_info({:fetched, text}, state) do
    state = put_in(state.text, text)

    {:noreply, state}
  end

  defp tick(state) do
    state =
      if state.position > 2 do
        state
        |> Map.put(:position, 0)
        |> Map.put(:letter, state.letter + 1)
      else
        Map.put(state, :position, state.position + 1)
      end

    if state.letter >= String.length(state.text) do
      Map.put(state, :letter, 0)
    else
      state
    end
  end

  defp draw_text(state, text, color, position, letter) do
    state
    |> Draw.char(get_letter(text, letter, 0),  0 - position, @offset, color)
    |> Draw.char(get_letter(text, letter, 1),  4 - position, @offset, color)
    |> Draw.char(get_letter(text, letter, 2),  8 - position, @offset, color)
    |> Draw.char(get_letter(text, letter, 3), 12 - position, @offset, color)
    |> Draw.char(get_letter(text, letter, 4), 16 - position, @offset, color)
  end

  defp get_letter(text, letter, pos) do
    len = String.length(text)

    letter =
      if letter + pos >= len do
        letter + pos - len
      else
        letter + pos
      end

    String.at(text, letter)
  end

  defp fetch do
    pid = self()

    spawn(fn ->
      try do
        send(pid, {:fetched, fetch_playing()})
      rescue
        _ -> :nothing
      end
    end)
  end

  def fetch_playing do
    data =
      "2.0/?method=user.getrecenttracks&user=#{Application.get_env(:lastfm, :username)}&api_key=#{Application.get_env(:lastfm, :api_key)}&format=json"
      |> get!()
      |> Map.get(:body)
      |> Map.get("recenttracks")
      |> Map.get("track")
      |> List.first()

    if data["@attr"]["nowplaying"] == "true" do
      artist =
        data
        |> Map.get("artist")
        |> Map.get("#text")
        |> String.downcase()

      song =
        data
        |> Map.get("name")
        |> String.downcase()

      "#{artist} #{song} "
    else
      "so... quiet... "
    end
  end
end

File: pix_esp8266/src/pix/screens/lastfm.cpp
#include "lastfm.h"
#include <cstdlib>
#include <cstring>

LastFM::LastFM(Platform *p) {
  platform = p;
  throttle = 4;
  refresh_every = 2 * 60;
  position = 16;
  refreshed_at = 0;
};

void LastFM::refresh() {
  std::string json =
      platform->fetch("https://ws.audioscrobbler.com/2.0/"
                      "?method=user.getrecenttracks"
                      "&limit=1&format=json"
                      "&user=" LASTFM_USER "&api_key=" LASTFM_KEY);

  if (json != "error") {
    JsonDocument doc;
    deserializeJson(doc, json);
    JsonVariant track = doc["recenttracks"]["track"][0];

    std::string name = track["name"].as<std::string>();
    std::string artist = track["artist"]["#text"].as<std::string>();
    now = track["@attr"]["nowplaying"];

    info = " " + artist + " - " + name;
  } else {
    info = "Error fetching data";
  }

  position = 16;
  len = info.length() * 4;
  screen_frames = len * throttle + 64;
};

void LastFM::update() {
  if (position < -len) {
    position = 16;
  }

  position--;

  platform->clear();

  if (now) {
    Chars::put_char(platform, 'N', 1, 0, YELLOW);
    Chars::put_char(platform, 'O', 6, 0, YELLOW);
    Chars::put_char(platform, 'W', 11, 0, YELLOW);
  } else {
    Chars::put_char(platform, 'L', 1, 0, GREEN);
    Chars::put_char(platform, 'A', 4, 0, GREEN);
    Chars::put_char(platform, 'S', 8, 0, GREEN);
    Chars::put_char(platform, 'T', 12, 0, GREEN);
  }

  for (int j = 0; j < 5; j++) {
    size_t i = j + (position / -4);
    if (i >= 0 && i < info.length()) {
      Chars::put_char(platform, info[i], (i * 4) + position, 8, BLUE);
    }
  }
};
